### PR TITLE
eliminate lodash library

### PR DESCRIPTION
**Pull request by jingt06**
#12 , removed lodash dependency.

### CODE REVIEW

File: src/ObjectGraph.es6.js [hunk 1]
```diff
@@ -96,8 +106,10 @@ ObjectGraph.prototype.blacklistedProperties = [
 
 ObjectGraph.prototype.initLazyData = function() {
   stdlib.memo(this, 'invTypes', () => {
-    let invTypes = remap['a:b=>b:[a]'](this.types);
-    _.forOwn(invTypes, (v, k, o) => o[k] = v[0]);
+    var invTypes = {};
+    Object.keys(this.types).map(type => {
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: Here and elsewhere: Have you benchmarked this? `for ( var key in o ) { if ( o.hasOwnProperty(key) ) ... }` generally performs faster, but if the difference is not appreciable in practice, I'm fine with this.

File: src/ObjectGraph.es6.js [hunk 2]
```diff
@@ -387,28 +399,27 @@ ObjectGraph.prototype.capture = function(o, opts) {
...
            ids.some(protoId => protoId === newProto))
       newProto = this.getPrototype(newProto);
 
-    _.forOwn(this.invProtos[id],
-             protoId => {
-               found = true;
-               console.assert(this.protos[protoId] === id);
-               this.protos[protoId] = newProto;
-             });
+    invProtoIds.forEach(invProtoId => {
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js:425]: Was there a data model change here? `_.forOwn` will iterate over properties and `a.forEach` will iterate over array elements, right?

**Review by mdittmer — COMMENTED** [review]
Just a few minor issues to discuss/address. Since we don't have tests setup for this repo, can you please do some substantial data collection on `master` and on `eliminate-lodash` to make sure the data are identical?

If that's already done, I trust you ;-)

File: src/ObjectGraph.es6.js (hunk 2 above)
Comment (jingt06) [reply src/ObjectGraph.es6.js:425]: this.invProtos returns {protoId: [list of object's id whose proto is protoId.]}. So _.forOwn(this.invProtos[id] is actually iterate over array elements, it should work the same as this.invProtos[id].forEach(...).

Comment (mdittmer) [reply src/ObjectGraph.es6.js:425]: Ah. You're right. Thanks for clarifying.

File: src/ObjectGraph.es6.js (hunk 1 above)
Comment (jingt06) [reply src/ObjectGraph.es6.js]: Done

File: src/ObjectGraph.es6.js [hunk 3]
```diff
@@ -398,15 +400,17 @@ ObjectGraph.prototype.capture = function(o, opts) {
...
-      var refIds = refKeys[refKey];
-      refIds.forEach(refId => {
-        found = true;
-        delete this.data[refId][refKey];
-      });
-    });
+  let invData = this.invData[id];
+  if ( invData ) {
+    for ( var key in invData ) {
+      if (invData.hasOwnProperty(key)){
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: nit: spacing

File: src/ObjectGraph.es6.js [hunk 4]
```diff
@@ -32,6 +30,18 @@ function ObjectGraph(opts) {
   this.init(opts);
 };
 
+function cloneDeep(obj) {
+  if (typeof obj !== 'object') {
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: nit: spacing.

File: src/ObjectGraph.es6.js [hunk 5]
```diff
@@ -386,29 +400,30 @@ ObjectGraph.prototype.capture = function(o, opts) {
...
   let found = false;
-  if ( this.invData[id] ) {
-    _.forOwn(
-      this.invData[id],
-      (refIds, refKey) => refIds.forEach(
-        refId => {
+  let invData = this.invData[id];
+  if ( invData ) {
+    for ( var key in invData ) {
+      if (invData.hasOwnProperty(key)){
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: nit: spacing

File: src/analysis.es6.js [hunk 6]
```diff
@@ -56,7 +54,15 @@ function intersectionIds1(g1, g2, match = anyAnyGraphMatcher) {
 }
 
 function differenceIds1(g1, g2, match = anyAnyGraphMatcher) {
-  return _.difference(g1.getAllIds(), intersectionIds1(g1, g2, match));
+  var diffArray = [];
+  var minuendArr = g1.getAllIds();
+  var subtrahendArr = intersectionIds1(g1, g2, match);
+  for (var i = 0; i < minuendArr.length; i ++) {
+    if (subtrahendArr.indexOf(minuendArr[i]) === -1) {
```
Comment (mdittmer) [inline src/analysis.es6.js]: nit: spacing

File: src/analysis.es6.js [hunk 7]
```diff
@@ -56,7 +54,15 @@ function intersectionIds1(g1, g2, match = anyAnyGraphMatcher) {
 }
 
 function differenceIds1(g1, g2, match = anyAnyGraphMatcher) {
-  return _.difference(g1.getAllIds(), intersectionIds1(g1, g2, match));
+  var diffArray = [];
+  var minuendArr = g1.getAllIds();
+  var subtrahendArr = intersectionIds1(g1, g2, match);
+  for (var i = 0; i < minuendArr.length; i ++) {
```
Comment (mdittmer) [inline src/analysis.es6.js]: nit: spacing

File: src/analysis.es6.js [hunk 8]
```diff
@@ -96,10 +102,8 @@ function intersectDifference(inGraphs, exGraphs, match = anyAnyGraphMatcher) {
   if (inGraphs.length === 1 && exGraphs.length === 0)
```
Comment (mdittmer) [inline src/analysis.es6.js]: nit: spacing

**Review by mdittmer — APPROVED** [review]
Identified a bunch of code style issues (including some that got by review previously). Whether or not you check it in, could you please create a `.eslintrc` that includes that extra spacing around `if`/`for`/`else` (this was inherited from FOAM's style guide and doesn't apply to all related projects in this space, as you know).

**Comment by jingt06** [comment]
I set up the eslint and add some rules based on the current style.

I don't think rules that includes that extra spacing around if/for/else exists in eslint.
There is a space-in-parens rule. However, when it is turned on, function calls must write as "functionaCall( args )", which does not follow the current style.

**Comment by jingt06** [comment]
Test is done using analyze_og and web catalog. I didn't see any differences compared to master branch.

File: .eslintrc [hunk 9]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (mdittmer) [inline .eslintrc]: If we can't get the most common error without a plugin, let's not check in config + deps on eslint.

File: src/ObjectGraph.es6.js [hunk 10]
```diff
@@ -98,18 +98,21 @@ ObjectGraph.prototype.blacklistedObjects = [
   value => value !== undefined
 );
 ObjectGraph.prototype.blacklistedProperties = [
-  ['MimeType', 'enabledPlugin'],
-    // MimeType's enabledPlugin will return a new MimeType object in Chrome
-    // (version<=42), which will cause a infinite recursion in the object graphs.
+  [ 'MimeType', 'enabledPlugin' ],
+    // MimeType's enabledPlugin will return a new MimeType object
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: nit: Let's indent this in-line with the `[`, and put it above the inner array (`[ 'MimeType', 'enabledPlugin' ]`).

**Review by mdittmer — APPROVED** [review]
A couple nits/things to drop from this Cl. LGTM!

File: src/ObjectGraph.es6.js (hunk 10 above)
Comment (jingt06) [reply src/ObjectGraph.es6.js]: Done

File: .eslintrc (hunk 9 above)
Comment (jingt06) [reply .eslintrc]: Done

**Pull request merged**